include/revng/Support/GraphAlgorithms.h

```diff
@@ -518,26 +541,844 @@ nodesBetweenNewImpl(GraphT Source,
 }
 
 template<class GraphT>
-inline llvm::SmallPtrSet<GraphT, 4>
+inline SmallSetVector<GraphT>
 nodesBetweenNew(GraphT Source,
                 GraphT Destination,
-                const llvm::SmallPtrSetImpl<GraphT> *IgnoreL = nullptr) {
+                const SmallSetVector<GraphT> *IgnoreList = nullptr) {
   return nodesBetweenNewImpl<GraphT, llvm::GraphTraits<GraphT>>(Source,
                                                                 Destination,
-                                                                IgnoreL);
+                                                                IgnoreList);
 }
 
 template<class GraphT>
-inline llvm::SmallPtrSet<GraphT, 4>
+inline SmallSetVector<GraphT>
 nodesBetweenNewReverse(GraphT Source,
                        GraphT Destination,
-                       const llvm::SmallPtrSetImpl<GraphT> *IgnoreL = nullptr) {
+                       const SmallSetVector<GraphT> *IgnoreList = nullptr) {
   using namespace llvm;
   return nodesBetweenNewImpl<GraphT, GraphTraits<Inverse<GraphT>>>(Source,
                                                                    Destination,
-                                                                   IgnoreL);
+                                                                   IgnoreList);
+}
+
+template<class NodeT>
+bool intersect(SmallSetVector<NodeT> &First, SmallSetVector<NodeT> &Second) {
+  // TODO: Upon migration to LLVM 17, we should use the `llvm::set_intersection`
+  // provided in `SetOperations.h` header file.
+  llvm::SmallPtrSet<NodeT, 4> Intersection(First.begin(), First.end());
+  llvm::set_intersect(Intersection, Second);
+  return !Intersection.empty();
+}
+
+template<class NodeT>
+bool subset(SmallSetVector<NodeT> &Contained,
+            SmallSetVector<NodeT> &Containing) {
+  return llvm::set_is_subset(Contained, Containing);
+}
+
+template<class NodeT>
+bool equal(SmallSetVector<NodeT> &First, SmallSetVector<NodeT> &Second) {
+  return llvm::set_is_subset(First, Second)
+         and llvm::set_is_subset(Second, First);
+}
```

Comment:
OK, but those are not `SmallSetVector`-specific.

I see no point in having a `subset` function which is just a wrapper around `set_is_subset`, just use `set_is_subset` directly.
I see the point for `equal` and `intersect`, but then do as `SetOperations.h` does (in fact, we can create our own `include/revng/ADT/SetOperations.h`) and make those function generic (i.e., not `SmallSetVector`-specific).

Additional note: `intersect` is a verb representing an action but the function actually returns a `bool`. Rename to `intersects`.

Reply:
You are right. In ab8422012f2a7aca302f31efb79cf14a80453de6 I added the new `include/revng/ADT/SetOperations.h` header, and defined `equal` and `intersect`, while I dropped `set_is_subset`.

The only thing is that, due to the internal functioning of `llvm::set_intersect` (there is some idiosyncrasy with how `llvm` handles the internal `erase` operation with `SmallSet`s like type) , there were some problems in adopting the template signature for `SetTy`. I think that the easiest solution is to make use of a new `llvm::set_intersection` function defined recently in the LLVM codebase, which does exactly what we want.

Therefore, I cherry-picked in https://github.com/revng/llvm-project/pull/28/commits/fd9d91de43f62a74bc9ba4e297389b7392adcad3, in the sister PR to the `llvm-project` repo, the, well isolated, llvm upstream commit that introduce it, so that we can adopt this approach without having to make hacks. Let me know if this seems reasonable to you.